flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/ScalarSqlFunction.scala

```diff
@@ -114,7 +114,9 @@ object ScalarSqlFunction {
 
         inferredTypes.zipWithIndex.foreach {
           case (inferredType, i) =>
-            operandTypes(i) = inferredType
+            if (operandTypes.length > 0) {
+              operandTypes(i) = inferredType
```

Comment:
If this is a varargs method, the inferredType is an array type. The operand type should be the component type of the array type, not the array type. 

And the `operandTypes.length > 0` condition is still not safe, say the method is `eval(String a, int... b)`  and calling `eval("hello")`, the `operandTypes`'s length is 1, but `inferredTypes`'s length is 2. An IndexOutOfBoundsException would be thrown as before.